astropy/io/ascii/qdp.py

```diff
@@ -0,0 +1,578 @@
+                      "table", AstropyUserWarning)
+        table_id = 0
+
+    tables = _get_tables_from_qdp_file(qdp_file, input_colnames=names, delimiter=delimiter)
+
+    return tables[table_id]
+
+
+def _write_table_qdp(table, filename=None, err_specs=None):
+    """Write a table to a QDP file
+
+    Parameters
+    ----------
+    table : :class:`~astropy.table.Table` object
+        Input table to be written
+    filename : str
+        Output QDP file name
+
+    Other parameters
+    ----------------
+    err_specs : dict
+        Dictionary of the format {'serr': [1], 'terr': [2, 3]}, specifying
+        which columns have symmetric and two-sided errors (see QDP format
+        specification)
+    """
+    import io
+    fobj = io.StringIO()
+
+    if 'initial_comments' in table.meta and table.meta['initial_comments'] != []:
+        for line in table.meta['initial_comments']:
+            line = line.strip()
+            if not line.startswith("!"):
+                line = "!" + line
+            print(line, file=fobj)
+
+    if err_specs is None:
+        serr_cols, terr_cols = _understand_err_col(table.colnames)
+    else:
+        serr_cols = err_specs.pop("serr", [])
+        terr_cols = err_specs.pop("terr", [])
+    if serr_cols != []:
+        col_string = " ".join([str(val) for val in serr_cols])
+        print(f"READ SERR {col_string}", file=fobj)
+    if terr_cols != []:
+        col_string = " ".join([str(val) for val in terr_cols])
+        print(f"READ TERR {col_string}", file=fobj)
+
+    if 'comments' in table.meta and table.meta['comments'] != []:
+        for line in table.meta['comments']:
+            line = line.strip()
+            if not line.startswith("!"):
+                line = "!" + line
+            print(line, file=fobj)
+
+    colnames = table.colnames
+    print("!" + " ".join(colnames), file=fobj)
+    for row in table:
+        values = []
+        for val in row:
+            if not np.ma.is_masked(val):
+                rep = str(val)
+            else:
+                rep = "NO"
+            values.append(rep)
+        print(" ".join(values), file=fobj)
+
+    full_string = fobj.getvalue()
+    fobj.close()
+
+    if filename is not None:
+        with open(filename, 'w') as fobj:
+            print(full_string, file=fobj)
+
+    return full_string.split("\n")
+
+
+class QDPSplitter(core.DefaultSplitter):
+    """
+    Split on space for QDP tables
+    """
+    delimiter = ' '
+
+
+class QDPHeader(basic.CommentedHeaderHeader):
+    """
+    Header that uses the :class:`astropy.io.ascii.basic.QDPSplitter`
+    """
+    splitter_class = QDPSplitter
+    comment = "!"
+    write_comment = "!"
+
+
+class QDPData(basic.BasicData):
+    """
+    Data that uses the :class:`astropy.io.ascii.basic.CsvSplitter`
+    """
+    splitter_class = QDPSplitter
+    fill_values = [(core.masked, 'NO')]
+    comment = "!"
+    write_comment = None
+
+
+class QDP(basic.Basic):
+    """QDP table.
```

Comment:
This main docstring should be expanded to illustrate the available parameters and usage. This will be the only way people know about the `header_id`, `err_specs` etc. For a good example see the following that shows a full example of the file format and how to read. You can also include a write example. https://github.com/astropy/astropy/blob/1ed1d945aadbdb9548e37eccfcc28115e3d03ca8/astropy/io/ascii/daophot.py#L315